Implement gt, gte, le, lte relational operators

workmajj opened the pull request:
I took a stab at implementing `>`, `>=`, `<`, and `<=` operators. These would be useful for checking, e.g., whether a given value is between two bounds when we're expressing field constraints.

Two questions about what makes sense for the language:

* When we compare two vanilla numbers, they're equal only if both their values and scales are equal. For example, `4.00` and `4.00` are equal, but `2.0` and `2.000` are not. For these new operators, I made the choice to relax the checks: I convert each number to the same scale, and then compare their values only. I did this because I think we'll use these operators for bounds checking, where strict equality (in the earlier sense) is not the goal. For example, having `5.9999 < 6.0` and `8.125 >= 3` both resolve to `true` seems important for this task. What do you think?

* Some other (existing) operators have explicit behavior when dealing with `undefined` values. I may still need to do that here. If we were writing a field constraint on `percent`, for example, we'd need to be able to express `percent > 0 && percent <= 0.06` while `percent` was still `undefined` (would at least want that whole thing to be `undefined`, rather than bomb out).

Thanks!

pascallouisperez commented on expressions.go:
```diff
@@ -186,6 +186,20 @@ func (e *tBinop) Compute(ws *Worksheet) (Value, error) {
 	}
 
+	// gt/gte/lt/lte
+	if e.op == opGT {
```
nicer as a switch

pascallouisperez commented on parser.go:
```diff
@@ -52,6 +52,10 @@ var (
 	pNot           = newTokenPattern("!", "\\!")
 	pEqual         = newTokenPattern("==", "\\=\\=")
 	pNotEqual      = newTokenPattern("!=", "\\!\\=")
+	pGT            = newTokenPattern(">", "\\>")
```
`pGt` for consistency

workmajj replied:
I believe the official guidance is to make initialisms either all lowercase, all uppercase--but never mixed:

> Words in names that are initialisms or acronyms (e.g. "URL" or "NATO") have a consistent case. For example, "URL" should appear as "URL" or "url" (as in "urlPony", or "URLPony"), never as "Url". As an example: ServeHTTP not ServeHttp. For identifiers with multiple initialized "words", use for example "xmlHTTPRequest" or "XMLHTTPRequest".

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

workmajj replied:
I think a good solution here is to change to `pGreaterThan`, etc., which is more literate anyway. :-)

pascallouisperez commented on parser_test.go:
```diff
@@ -213,6 +213,140 @@ func (s *Zuite) TestParser_parseExpressionsAndCheckCompute() {
 		`false && undefined`:               `false`,
 		`false && 6 / 0 round down 7 == 6`: `false`,
 
+		// gt
```
that's a lot of tests... fine leaving as is, but generally having a smaller set of unique examples is easier to read

workmajj replied:
I originally made these in a pretty systematic way, even though there are a lot. There were a bunch of things I wanted to check (precedence, comparisons with different scales, etc.), and then I wanted 3x tests for each (whether the left was greater, right was greater, or the sides were equal).

I think there are some I can de-dup, so I'll do that, but I don't want you to think I was nuts. 😆

pascallouisperez commented on parser_test.go:
```diff
@@ -213,6 +213,140 @@ func (s *Zuite) TestParser_parseExpressionsAndCheckCompute() {
+		`-1.5000 <= -1.5`: `true`,
+
+		// more complicated gt/gte/lt/lte expressions
+		`15.899 > 15 + 0.899 round up 0`:        `false`,
```
maybe more interesting `15.899 > 15 + 0.8999 round down 3`

pascallouisperez commented on tree.go:
```diff
@@ -62,6 +62,10 @@ const (
 	opNot          = "not"
 	opEqual        = "equal"
 	opNotEqual     = "not-equal"
+	opGT           = "gt"
```
`opGt` for consistency

pascallouisperez commented on values.go:
```diff
@@ -116,6 +116,8 @@ func (value *Undefined) Equal(that Value) bool {
 	return ok
 }
 
+// TODO: jjw: implement gte/lte for undefined for consistency with eq?
```
yes! `undefined < X` would be `undefined`

pascallouisperez commented on values.go:
```diff
@@ -124,6 +126,7 @@ func (value *Number) Type() Type {
 	return value.typ
 }
 
+// usual number equality requires that both value and scale are equal
```
which isn't what we want in the language when doing `==` because we'd very likely do `a != 0 && b / a` and would not want that to be a division by zero

we should fix equality separately to focus on numerical equality

on style: comment in Go should start by the function name e.g. `Equal compares two ....`

pascallouisperez commented on values.go:
```diff
@@ -132,6 +135,61 @@ func (value *Number) Equal(that Value) bool {
+// for usual number equality, we require that value and scale are equal;
+// for gt/gte/lt/lte comparisons, however, we recognize that scale doesn't need
+// to be equal because we'll often be bounds checking (e.g., 5.999999 <= 6.0)
+func (value *Number) ScaledEqual(that Value) bool {
```
Maybe `NumericEqual` and make it clear that `that` can only be a number. Here we can ask the caller to check that the value is a number,

pascallouisperez commented on values.go:
```diff
@@ -132,6 +135,61 @@ func (value *Number) Equal(that Value) bool {
 	return value.value == typed.value && value.typ.scale == typed.typ.scale
 }
 
+func (value *Number) GreaterThan(that Value) bool {
```
Let's only have `GreaterThan` defined on `that *Number` similar to other operations like `Plus`.

In the `Binop` compute you can handle with non-numerical cases. You don't want to do this at the value level.

pascallouisperez commented on values.go:
```diff
@@ -132,6 +135,61 @@ func (value *Number) Equal(that Value) bool {
+	return value.value == typed.value
+}
+
+func (value *Number) GreaterThanOrEqual(that Value) bool {
```
ditto on type of `that`

pascallouisperez commented on values.go:
```diff
@@ -132,6 +135,61 @@ func (value *Number) Equal(that Value) bool {
+	return value.ScaledEqual(typed) || value.GreaterThan(typed)
+}
+
+func (value *Number) LessThan(that Value) bool {
```
ditto on type of `that`

pascallouisperez commented on values.go:
```diff
@@ -132,6 +135,61 @@ func (value *Number) Equal(that Value) bool {
+	return !value.ScaledEqual(typed) && !value.GreaterThan(typed)
+}
+
+func (value *Number) LessThanOrEqual(that Value) bool {
```
ditto on type of `that`

workmajj commented:
@pascallouisperez I think I've addressed all your comments here!

I'll make a separate issue to change the way we do numerical equality, per your note.
